core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -811,40 +814,53 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         }
     }
 
-    private static final int[] EMPTY_DOC_IDS = new int[0];
-
     /**
      * Shortcut ids to load, we load only "from" and up to "size". The phase controller
      * handles this as well since the result is always size * shards for Q_A_F
      */
     private void shortcutDocIdsToLoad(SearchContext context) {
+        final int[] searchDocIdsToLoad;
         if (context.request().scroll() != null) {
             TopDocs topDocs = context.queryResult().topDocs();
-            int[] docIdsToLoad = new int[topDocs.scoreDocs.length];
+            searchDocIdsToLoad = new int[topDocs.scoreDocs.length];
             for (int i = 0; i < topDocs.scoreDocs.length; i++) {
-                docIdsToLoad[i] = topDocs.scoreDocs[i].doc;
+                searchDocIdsToLoad[i] = topDocs.scoreDocs[i].doc;
             }
-            context.docIdsToLoad(docIdsToLoad, 0, docIdsToLoad.length);
         } else {
             TopDocs topDocs = context.queryResult().topDocs();
             if (topDocs.scoreDocs.length < context.from()) {
                 // no more docs...
-                context.docIdsToLoad(EMPTY_DOC_IDS, 0, 0);
-                return;
-            }
-            int totalSize = context.from() + context.size();
-            int[] docIdsToLoad = new int[Math.min(topDocs.scoreDocs.length - context.from(), context.size())];
-            int counter = 0;
-            for (int i = context.from(); i < totalSize; i++) {
-                if (i < topDocs.scoreDocs.length) {
-                    docIdsToLoad[counter] = topDocs.scoreDocs[i].doc;
-                } else {
-                    break;
+                searchDocIdsToLoad = new int[0];
+            } else {
+                int totalSize = context.from() + context.size();
+                searchDocIdsToLoad = new int[Math.min(topDocs.scoreDocs.length - context.from(), context.size())];
+                for (int i = context.from(); i < Math.min(totalSize, topDocs.scoreDocs.length); i++) {
+                    searchDocIdsToLoad[i - context.from()] = topDocs.scoreDocs[i].doc;
                 }
-                counter++;
             }
-            context.docIdsToLoad(docIdsToLoad, 0, counter);
         }
+        int[] docIdsToLoad = searchDocIdsToLoad;
+        final Suggest suggest = context.queryResult().suggest();
+        if (suggest != null && suggest.hasScoreDocs()) {
+            int numSuggestDocs = 0;
+            for (CompletionSuggestion completionSuggestion : suggest.filter(CompletionSuggestion.class)) {
+                numSuggestDocs += completionSuggestion.getEntries().get(0).getOptions().size();
+            }
+            docIdsToLoad = new int[searchDocIdsToLoad.length + numSuggestDocs];
+            System.arraycopy(searchDocIdsToLoad, 0, docIdsToLoad, 0, searchDocIdsToLoad.length);
+            int offset = searchDocIdsToLoad.length;
+            for (CompletionSuggestion completionSuggestion : suggest.filter(CompletionSuggestion.class)) {
```

Comment:
this code and the code in `SearchPhaseController#sortDocs` is almost identical. I think the only difference is the assignment of the shard ID. Maybe we can factor this out into a static method and use it in both places. It would be good to redcue the duplication on such a level and it would increase the test coverage. I would be in favor of that.